// semver/index.js
'use strict'

// just pre-load all the stuff that index.js lazily exports
const internalRe = require('./internal/re')
// Import module dependency.
const constants = require('./internal/constants')
// Import module dependency.
const SemVer = require('./classes/semver')
// Import module dependency.
const identifiers = require('./internal/identifiers')
// Import module dependency.
const parse = require('./functions/parse')
// Import module dependency.
const valid = require('./functions/valid')
// Import module dependency.
const clean = require('./functions/clean')
// Import module dependency.
const inc = require('./functions/inc')
// Import module dependency.
const diff = require('./functions/diff')
// Import module dependency.
const major = require('./functions/major')
// Import module dependency.
const minor = require('./functions/minor')
// Import module dependency.
const patch = require('./functions/patch')
// Import module dependency.
const prerelease = require('./functions/prerelease')
// Import module dependency.
const compare = require('./functions/compare')
// Import module dependency.
const rcompare = require('./functions/rcompare')
// Import module dependency.
const compareLoose = require('./functions/compare-loose')
// Import module dependency.
const compareBuild = require('./functions/compare-build')
// Import module dependency.
const sort = require('./functions/sort')
// Import module dependency.
const rsort = require('./functions/rsort')
// Import module dependency.
const gt = require('./functions/gt')
// Import module dependency.
const lt = require('./functions/lt')
// Import module dependency.
const eq = require('./functions/eq')
// Import module dependency.
const neq = require('./functions/neq')
// Import module dependency.
const gte = require('./functions/gte')
// Import module dependency.
const lte = require('./functions/lte')
// Import module dependency.
const cmp = require('./functions/cmp')
// Import module dependency.
const coerce = require('./functions/coerce')
// Import module dependency.
const truncate = require('./functions/truncate')
// Import module dependency.
const Comparator = require('./classes/comparator')
// Import module dependency.
const Range = require('./classes/range')
// Import module dependency.
const satisfies = require('./functions/satisfies')
// Import module dependency.
const toComparators = require('./ranges/to-comparators')
// Import module dependency.
const maxSatisfying = require('./ranges/max-satisfying')
// Import module dependency.
const minSatisfying = require('./ranges/min-satisfying')
// Import module dependency.
const minVersion = require('./ranges/min-version')
// Import module dependency.
const validRange = require('./ranges/valid')
// Import module dependency.
const outside = require('./ranges/outside')
// Import module dependency.
const gtr = require('./ranges/gtr')
// Import module dependency.
const ltr = require('./ranges/ltr')
// Import module dependency.
const intersects = require('./ranges/intersects')
// Import module dependency.
const simplifyRange = require('./ranges/simplify')
// Import module dependency.
const subset = require('./ranges/subset')
module.exports = {
  parse,
  valid,
  clean,
  inc,
  diff,
  major,
  minor,
  patch,
  prerelease,
  compare,
  rcompare,
  compareLoose,
  compareBuild,
  sort,
  rsort,
  gt,
  lt,
  eq,
  neq,
  gte,
  lte,
  cmp,
  coerce,
  truncate,
  Comparator,
  Range,
  satisfies,
  toComparators,
  maxSatisfying,
  minSatisfying,
  minVersion,
  validRange,
  outside,
  gtr,
  ltr,
  intersects,
  simplifyRange,
  subset,
  SemVer,
  re: internalRe.re,
  src: internalRe.src,
  tokens: internalRe.t,
  SEMVER_SPEC_VERSION: constants.SEMVER_SPEC_VERSION,
  RELEASE_TYPES: constants.RELEASE_TYPES,
  compareIdentifiers: identifiers.compareIdentifiers,
  rcompareIdentifiers: identifiers.rcompareIdentifiers,
}
