// mongodb/lib/operations/end_sessions.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.EndSessionsOperation = void 0;
// Import module dependency.
const responses_1 = require("../cmap/wire_protocol/responses");
// Import module dependency.
const command_1 = require("../operations/command");
// Import module dependency.
const read_preference_1 = require("../read_preference");
// Import module dependency.
const utils_1 = require("../utils");
// Import module dependency.
const operation_1 = require("./operation");
class EndSessionsOperation extends command_1.CommandOperation {
    constructor(sessions) {
        super();
        this.writeConcern = { w: 0 };
        this.ns = utils_1.MongoDBNamespace.fromString('admin.$cmd');
        this.SERVER_COMMAND_RESPONSE_TYPE = responses_1.MongoDBResponse;
        this.sessions = sessions;
    }
    buildCommandDocument(_connection, _session) {
        return {
            endSessions: this.sessions
        };
    }
    buildOptions(timeoutContext) {
        return {
            timeoutContext,
            readPreference: read_preference_1.ReadPreference.primaryPreferred
        };
    }
    get commandName() {
        return 'endSessions';
    }
}
exports.EndSessionsOperation = EndSessionsOperation;
(0, operation_1.defineAspects)(EndSessionsOperation, operation_1.Aspect.WRITE_OPERATION);
